src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs

```diff
@@ -276,12 +279,149 @@ internal unsafe SocketError DoOperationSend(SafeCloseSocket handle)
+                }
+
+                // Windows differentiates the directory not being found from the file not being found.
+                // Approximate this by checking to see if the directory exists; this is only best-effort,
+                // as there are various things that could affect this, e.g. directory creation racing with
+                // this check, but it's good enough for most situations.
+                if (exc is FileNotFoundException fnfe)
+                {
+                    string dirname = Path.GetDirectoryName(fnfe.FileName);
+                    if (!string.IsNullOrEmpty(dirname) && !Directory.Exists(dirname))
+                    {
+                        throw new DirectoryNotFoundException(fnfe.Message);
+                    }
+                }
+
+                // Otherwise propagate the original error.
+                throw;
+            }
+
+            SendPacketsAsync(socket, elements, files);
+            return SocketError.IOPending;
+        }
+
+        private async void SendPacketsAsync(Socket socket, SendPacketsElement[] elements, FileStream[] files) // async void, as completion is communicated via other means
+        {
+            try
+            {
+                int bytesTransferred = 0;
+                Debug.Assert(elements.Length == files.Length);
+                for (int i = 0; i < elements.Length; i++)
+                {
+                    SendPacketsElement e = elements[i];
+                    if (e != null)
+                    {
+                        if (e.FilePath == null)
+                        {
+                            bytesTransferred += await socket.SendAsync(new ArraySegment<byte>(e.Buffer, e.Offset, e.Count), SocketFlags.None).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            FileStream fs = files[i];
+                            Debug.Assert(fs != null);
+                            using (fs)
+                            {
+                                // Validate that the offset and count are within the file.
+                                // If the count is 0, that means to send the whole file.
+                                long fileLen = fs.Length;
+                                if (e.Count > 0 && e.Offset > fs.Length - e.Count)
+                                {
+                                    throw new ArgumentOutOfRangeException();
+                                }
+
+                                // If there's an offset, move forward to that position.
+                                if (e.Offset != 0)
+                                {
+                                    fs.Position = e.Offset;
+                                }
+                                
+                                // Copy the requested number of bytes from the file to the socket.
+                                byte[] buffer = ArrayPool<byte>.Shared.Rent(0x1000);
```

Comment:
Wouldn't that significantly complicate the code? It'd be easy if I could use the public BeginSendFile, but that a) only works with paths, and b) doesn't support offset/count. To use sendfile directly, I'd need to plumb this all through all of the async infrastructure, no?